File: c/armstrong-numbers/armstrong_numbers.c
#include "armstrong_numbers.h"

#include <math.h>

bool is_armstrong_number(int candidate) {
  int digits[64];
  int i = 0;
  int n = candidate;
  while (n > 0) {
    digits[i] = n % 10;
    n /= 10;
    i += 1;
  }

  int sum = 0;
  for (int j = 0; j < i; ++j) {
    sum += pow(digits[j], i);
  }
  return sum == candidate;
}

File: zig/armstrong-numbers/armstrong_numbers.zig
const std = @import("std");

pub fn isArmstrongNumber(num: u128) bool {
    var cursor: usize = 0;
    var digits: [64]u4 = undefined;
    var n = num;
    while (n > 0) : (n /= 10) {
        digits[cursor] = @intCast(n % 10);
        cursor += 1;
    }

    var sum: u128 = 0;
    for (0..cursor) |i| {
        sum += std.math.powi(u128, digits[i], cursor) catch unreachable;
    }
    return sum == num;
}
